controllers/datadogagent/controller_reconcile_agent.go

```diff
@@ -283,30 +283,45 @@ func (r *Reconciler) labelNodesWithProfiles(ctx context.Context, profilesByNode
 			return err
 		}
 
-		_, profileLabelExists := node.Labels[agentprofile.ProfileLabelKey]
-
 		newLabels := map[string]string{}
-		for k, v := range node.Labels {
-			// If the profile uses the old profile label key, it should be removed
-			if k != agentprofile.OldProfileLabelKey {
-				newLabels[k] = v
-			}
-		}
+		labelsToRemove := map[string]bool{}
+		labelsToAddOrChange := map[string]string{}
 
 		// If the profile is the default one and the label exists in the node,
 		// it should be removed.
-		if isDefaultProfile && profileLabelExists {
-			delete(newLabels, agentprofile.ProfileLabelKey)
+		if isDefaultProfile {
+			if _, profileLabelExists := node.Labels[agentprofile.ProfileLabelKey]; profileLabelExists {
+				labelsToRemove[agentprofile.ProfileLabelKey] = true
+			}
+		} else {
+			// If the profile is not the default one and the label does not exist in
+			// the node, it should be added. If the label value is outdated, it
+			// should be updated.
+			if profileLabelValue := node.Labels[agentprofile.ProfileLabelKey]; profileLabelValue != profileNamespacedName.Name {
+				labelsToAddOrChange[agentprofile.ProfileLabelKey] = profileNamespacedName.Name
+			}
+		}
+
+		// Remove old profile label key if it is present
+		if _, oldProfileLabelExists := node.Labels[agentprofile.OldProfileLabelKey]; oldProfileLabelExists {
+			labelsToRemove[agentprofile.OldProfileLabelKey] = true
 		}
 
-		// If the profile is not the default one and the label does not exist in
-		// the node, it should be added.
-		if !isDefaultProfile && !profileLabelExists {
-			newLabels[agentprofile.ProfileLabelKey] = profileNamespacedName.Name
+		if len(labelsToRemove) > 0 || len(labelsToAddOrChange) > 0 {
```

Comment:
is there a benefit to having this `||` instead of two separate conditionals?

Reply:
The first for loop for `labelsToRemove` goes through all the node labels but the second one for `labelsToAddOrChange` doesn't. If we separated them, we'd need to loop through `node.Labels` twice

Reply:
ah i see, thanks!